menu.py

```diff
@@ -102,100 +109,100 @@ def setup_menu(self, surface, background_colour=BLUE):
-                    logger.debug("screen pressed: %s", event.pos)
-                    pos = (event.pos[0], event.pos[1])
-                    # for debugging purposes - adds a small dot
-                    # where the screen is pressed
-                    # pygame.draw.circle(screen, WHITE, pos, 2, 0)
-                    requested_challenge = self.on_click(pos)
-                    logger.info("requested challenge is %s", requested_challenge)
-                    if requested_challenge:
-                        logger.info("about to stop a thread if there's one running")
-                        if running_challenge:
-                            logger.info("about to stop thread")
-                            self.stop_threads(running_challenge)
-
-                    if requested_challenge is not None and requested_challenge is not "Exit" and requested_challenge is not "Other":
-                        running_challenge = self.launch_challenge(requested_challenge)
-                        logger.info("challenge %s launched", running_challenge.name)
-
-                    if requested_challenge == "Exit":
-                        sys.exit()
-                # ensure there is always a safe way to end the program
-                # if the touch screen fails
-                if event.type == KEYDOWN:
-                    if event.key == K_ESCAPE:
-                        sys.exit()
+        with ControllerResource() as self.joystick:
+            while True:
+                time = clock.tick(30)
+                pygame.display.update()
+                sgc.update(time)
+                if self.joystick.connected:
+                    #print("joystick connected")
+                    self.joystick_handler(self.joystick.check_presses())
+                for event in pygame.event.get():
+                    sgc.event(event)
+                    if event.type== GUI:
+                        if event.widget_type is "Button":
+                            requested_challenge = self.button_handler(event)
+                            if requested_challenge:
+                                logger.info("about to stop a thread if there's one running")
+                                if running_challenge:
+                                    logger.info("about to stop thread")
+                                    self.stop_threads(running_challenge)
+                            if requested_challenge is not None and requested_challenge is not "Exit" and requested_challenge is not "Other":
+                                running_challenge = self.launch_challenge(requested_challenge)
+                                logger.info("challenge %s launched", running_challenge.name)
+                            if requested_challenge == "Exit":
+                                sys.exit()
+                    # ensure there is always a safe way to end the program
+                    # if the touch screen fails
+                    if event.type == KEYDOWN:
```

Comment:
`if event.type == KEYDOWN` and event.key == K_ESCAPE`?

or just if `event.key == K_ESCAPE`?

Reply:
do all events have a key property?